src/plugins/lexer/mod.ts

```diff
@@ -0,0 +1,390 @@
+            break;
+          }
+
+          default: {
+            // exists for debugging purpose. this error should never occur in normal use
+            if (this.token.length > 4) {
+              throw new Error(
+                `Unexpected compiler token bypassed 4 character limit: ${this.token}`,
+              );
+            } else {
+              return;
+            }
+          }
+        }
+        this.mode = Mode.Token;
+        break;
+      }
+    }
+  }
+
+  // increment ahead `x` tokens
+  private next(jump: number = 1) {
+    this.virtual_position += jump;
+    return (this.source[this.position += jump]);
+  }
+
+  // decrement back `x` tokens
+  private previous(jump: number = 1) {
+    if (this.position === 0) return;
+    this.virtual_position -= jump;
+    return (this.source[this.position -= jump]);
+  }
+
+  // look ahead `x` tokens
+  private peek(distance: number = 1) {
+    if (this.position >= this.source.length) return " ";
+    return this.source[this.position + distance];
+  }
+
+  // compare comprehended token to a list of possible values
+  private checkToken(offset: number = 0) {
+    // sets end column (mainly for pretty error handling in parser)
+    this.trace.column[1] = this.virtual_position - 1 + offset;
+    let type: Token["type"];
+    const token = this.token;
+    switch (token) {
+      case "const":
+      case "mut":
+      case "if":
+      case "else":
+      case "while":
+      case "loop":
+      case "true":
+      case "false":
+      case "expose": {
+        type = "keyw";
+        break;
+      }
+      case "string":
+      case "number":
+      case "boolean":
+      case "collection":
+      case "unsafe": {
+        type = "type";
+        break;
+      }
+      case "+":
+        type = "add";
+        break;
+      case "-":
+        type = "sub";
+        break;
+      case "*":
+        type = "mult";
+        break;
+      case "/":
+        type = "div";
+        break;
+      case "%":
+        type = "mod";
+        break;
+      case "^":
+        type = "power";
+        break;
+      case "!":
+        type = "not";
+        break;
+      case "|":
+        type = "pipe";
+        break;
+      case ",":
+        type = "comma";
+        break;
+      case ":": {
+        type = "colon";
+        break;
+      }
+      case "=": {
+        type = "eq";
+        break;
+      }
+      case ">": {
+        type = "gt";
+        break;
+      }
+      case "<": {
+        type = "lt";
+        break;
+      }
+      case "{": {
+        type = "lcb";
+        break;
+      }
+      case "}": {
+        type = "rcb";
+        break;
+      }
+      case "(": {
+        type = "lp";
+        break;
+      }
+      case ")": {
+        type = "rp";
+        break;
+      }
+      case "": {
+        return;
+      }
+      default: {
+        if (isNumber(token)) {
+          type = "number";
+        } else {
+          type = "ident";
```

Comment:
This isn't necessarily true. We can't assume any token value is an identifier just because it doesn't reach any other case. The way I handled this in the legacy lexer was by checking if the previous token was a `const` or `mut` keyword.

This method is effective because when defining an identifier, we require either of these keywords; and when calling an identifier we use `@`, so we can assume *every* identifier token will follow a `keyw `token with a `const` or `mut` value.

Reply:
As for your suggestion to token checking, much of that is better left deferred to parser; lexer is naive in nature and only tasked with understanding what each token means, with more detailed analysis left up the parser. Having this check also ends up repetitive, as the end goal is a concurrent compilation system; there's no performance drawback from leaving until parser stage if each token is individually sent.

As for checking references, definitely something I missed in token analysis; will have lexer actively check for `@` symbol and break apart chained references (such as `@std.write()` for example).

Reply:
I think you missed the point I was making with this comment. I'm not saying we should care about the check for the purpose of analysis, but rather for the purpose of being correct. Right now the lexer will say *anything* is an identifier simply when it doesn't know what it is. Previously we had an unknown token type, and I recommend bringing this back. Here's an example: `const x: string { "hey" bug };`

This line has two syntax errors: `bug` & `;`. We know `bug` isn't a call to a variable since it doesn't have an `@`, and the semicolon simply doesn't exist. These are both invalid segments within the code -- they don't have a valid token type because they shouldn't be there. Rather than realizing this fact, this PR will assume those are identifiers... which doesn't make logical sense. An ident is strictly the name of a variable, function, or other user-generated names -- not a random piece of text within the code. 

So my original solution for this was to check if it came after a `const` or `mut`. This was effective because it made sure any unknown text that came after a defining keyword was known as an ident -- I really didn't care to analyze what type it was or any of that, I only cared about being correct when calling it an ident. If you have a simpler/faster method for accomplishing this, I'm all for it, but It's important that token types are actually correct. The assumption that any unknown code is an identifier is simply incorrect, so I think a token for unknown/invalid code would be an appropriate addition (& would help with errors in the parser).

Reply:
For token types, I think the best approach is to change the name from "ident" to "unknown", as variable names will be understood by the parser fairly easily. An individual token specific to variable identifiers would prove useless to parser.

if you're good with that, I can do the variable change :+1:

And yeah, references is definitely an important part to miss :sweat_smile:

Reply:
Yeah, this is a good approach. Any code segment that isn't known to the lexer, be it an identifier name or some random piece of text, will be an unknown token. The parser can then do the analysis between if it came after a `const` or `mut`, or if it's truly an unknown piece of text (i.e syntax error).

Good compromise -- Add this in your next commit & mark as resolved 👍